### PR TITLE
#1538 - Activity During Date Range Metric

### PR DESCRIPTION
**Pull request by kodinkat**
- fixes: #1538 
- fixes: #2080 
---

- New Activity During Date Range project metric introduced.
- Metric provides the ability to query post records by field and within a given date range.
- Ability to click-through to listed results post records for more detail.
- Added 2nd level filter support for typeahead based field types: `connection`, `location` & `user_select`.
- Display new value field was updated with.

Screenshot 2023-05-30 at 16 45 31

Screenshot 2023-05-30 at 16 46 19

Screenshot 2023-05-30 at 16 47 06

### CODE REVIEW

**Comment by corsacca** [comment]
@kodinkat 
Enable first layer all fields
image

List will show field changes during the time period.
ex all contact with baptism date change in range
ex all contact the status change in range

Some fields have optional second filter ( all contacts set to status active. all contacts assigned to user y)
-  key select, tags, multisilect, connection have

Please test and confirm fields first lever:

- [ ] multi_select
- [ ] key_select
- [ ] tags
- [ ] communication_channel
- [ ] connection
- [ ] user_select
- [ ] text
- [ ] textarea
- [ ] link
- [ ] date
- [ ] number
- [ ] boolean
- [ ] location
- [ ] location_meta

With optional second filter:
- [ ] multi_select
- [ ] key_select
- [ ] tags
- [ ] communication_channel
- [ ] connection
- [ ] user_select
- [ ] location?

**Comment by kodinkat** [comment]
@corsacca Ready for review...

**Comment by corsacca** [comment]
Error on Baptism Date all time
image

Connection fields seem to be returning values for any connection field.

**Comment by kodinkat** [comment]
@corsacca hopefully all major bugs now smoked out, but let me know if any more are encountered....

**Comment by corsacca** [comment]
@kodinkat 

- can you try adding the demo data from the demo plugin? The activity from connections fields from it don't show.
- Adding and then removing a connection should not show the same thing twice:
image

File: dt-metrics/combined/date-range-activity.php [hunk 1]
```diff
@@ -0,0 +1,370 @@
...
+
+    public function field_settings( WP_REST_Request $request ) {
+        if ( !$this->has_permission() ) {
+            wp_send_json_error( new WP_Error( 'get_field_settings', 'Missing Permissions', [ 'status' => 400 ] ) );
+        }
+        $url_params = $request->get_url_params();
+        return $this->get_field_settings( $url_params['post_type'] );
+    }
+
+    public function render_field_html( WP_REST_Request $request ){
```
Comment (corsacca) [inline dt-metrics/combined/date-range-activity.php:176]: @kodinkat how can we steer away from sending html in API calls? 
It's a security risk as the result can't be escaped

File: dt-metrics/combined/date-range-activity.php [hunk 2]
```diff
@@ -0,0 +1,370 @@
...
+                $meta_key_sql = ( $field_type == 'connection' ) ? "AND meta_key LIKE '%'" : "AND meta_key LIKE '" . esc_sql( $params['field'] ) . "'";
+
+                $values = [];
+                foreach ( $params['value'] ?? [] as $value ){
+                    if ( isset( $value['ID'] ) ){
+                        $values[] = $value['ID'];
+                    }
+                }
+                $meta_value_sql = ( !empty( $values ) ? 'AND meta_value IN (' . dt_array_to_sql( $values ) . ')' : "AND meta_value LIKE '%'" );
+                $obj_subtype_sql = ( $field_type == 'connection' ) ? "AND object_subtype = '" . esc_sql( $params['field'] ) . "'" : '';
```
Comment (corsacca) [inline dt-metrics/combined/date-range-activity.php:239]: @kodinkat object subtype used to be 'p2p', so this does not pick up on historical data.
image

**Review by corsacca — CHANGES_REQUESTED** [review]

File: dt-metrics/combined/date-range-activity.php [hunk 3]
```diff
@@ -0,0 +1,370 @@
...
+                [
+                    'methods'  => WP_REST_Server::READABLE,
+                    'callback' => [ $this, 'render_field_html' ],
+                    'permission_callback' => '__return_true'
+                ]
+            ]
+        );
+
+        register_rest_route(
+            $namespace, '/metrics/dummy_endpoint', [
```
[the remark this answers is not in the document]
Comment (kodinkat) [reply dt-metrics/combined/date-range-activity.php:152]: Given `window.METRICS.setupDatePicker()` shape, `dummy_endpoint` acts as a placeholder; used when initialising date range picker element.

File: dt-metrics/combined/date-range-activity.php (hunk 1 above)
Comment (kodinkat) [reply dt-metrics/combined/date-range-activity.php:181]: Could capture all field html when initially loading the page; or, refresh page and load on refresh... Personally, I'm not keen on either approach... Alternatively, html could be generated within frontend, but this too would add to code / logic duplication.

File: dt-metrics/combined/date-range-activity.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (corsacca) [reply dt-metrics/combined/date-range-activity.php:152]: @kodinkat what about the existing window.METRICS.setupDatePickerWithoutEndpoint() ?

File: dt-metrics/combined/date-range-activity.php (hunk 2 above)
Comment (kodinkat) [reply dt-metrics/combined/date-range-activity.php]: Historical `p2p` activity entries now captured; however, please note that any `p2p` entries detected within given date range, will be included; as there is no apparent way to filter by field id.

File: dt-metrics/combined/date-range-activity.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (kodinkat) [reply dt-metrics/combined/date-range-activity.php:152]: ah, good spot; as I did not see that! :)

File: dt-metrics/combined/date-range-activity.php (hunk 2 above)
Comment (corsacca) [reply dt-metrics/combined/date-range-activity.php]: @kodinkat I think there are 2options here
We use the meta_key and the field type:
image


Or we write a migration to upgrade the old activity to match the new shape.

Comment (kodinkat) [reply dt-metrics/combined/date-range-activity.php]: @corsacca yep, maybe the first option, re exploring the use of both the meta_key and field_type, to determine actual field_id.... As the 2nd option sounds like a new ticket? ;)

Comment (corsacca) [reply dt-metrics/combined/date-range-activity.php]: Yeah. Let's go for #1 and see where we get.

Comment (kodinkat) [reply dt-metrics/combined/date-range-activity.php]: @corsacca does the latest sql tweak hold true for you?

Comment (corsacca) [reply dt-metrics/combined/date-range-activity.php]: @kodinkat almost. 
We do need to add the field type. Other wise it gets both the  `Baptized` and the `baptized by` field.
image

Safir:

**Comment by corsacca** [comment]
Thank you @kodinkat !

**Pull request merged**